routing/registry.go

```diff
@@ -0,0 +1,142 @@
+package routing
+
+import (
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type metrics interface {
+	DetectedTime() time.Time
+	InflightRequests() int64
+}
+
+type entry struct {
+	detected         time.Time
+	inflightRequests int64
+}
+
+var _ metrics = &entry{}
+
+func (e *entry) DetectedTime() time.Time {
+	return e.detected
+}
+
+func (e *entry) InflightRequests() int64 {
+	return e.inflightRequests
+}
+
+type Registry struct {
+	mu sync.Mutex
+
+	data map[string]*entry
+}
+
+var _ PostProcessor = &Registry{}
+
+type RegistryOptions struct {
+}
+
+func normalizeSchemeHost(s, h string) (string, string, error) {
+	// endpoint address cannot contain path, the rest is not case sensitive
+	s, h = strings.ToLower(s), strings.ToLower(h)
+
+	hh, p, err := net.SplitHostPort(h)
+	if err != nil {
+		// what is the actual right way of doing this, considering IPv6 addresses, too?
+		if !strings.Contains(err.Error(), "missing port") {
+			return "", "", err
+		}
+
+		p = ""
+	} else {
+		h = hh
+	}
+
+	switch {
+	case p == "" && s == "http":
+		p = "80"
+	case p == "" && s == "https":
+		p = "443"
+	}
+
+	h = net.JoinHostPort(h, p)
+	return s, h, nil
+}
+
+func (r *Registry) Do(routes []*Route) []*Route {
+	const configErrFmt = "Error while processing endpoint fade-in settings: %s, %s, %v."
+	for _, route := range routes {
+		for i := range route.LBEndpoints {
+			epi := &route.LBEndpoints[i]
+			s, h, err := normalizeSchemeHost(epi.Scheme, epi.Host)
+			if err != nil {
+				log.Errorf(configErrFmt, epi.Scheme, epi.Host, err)
+				continue
+			}
+
+			epi.Scheme = s
```

Comment:
It's a good idea to normalize scheme+host here. However, it would be better to normalize them at the very creation of LBEndpoint structure.